### PR TITLE
API documentation.

### PR DESCRIPTION
**Pull request by jamadden**
This will allow for intersphinx links to work from other projects.

This may need to have the RTD box "install project in virtual env"
checked if it isn't already.

The version number in the docs is out of sync, I'll address that in a follow-up PR.

### CODE REVIEW

**Comment by jamadden** [comment]
See https://github.com/zopefoundation/z3c.schema/pull/11#issuecomment-247008485

**Comment by mgedmin** [comment]
LGTM.

**Comment by jamadden** [comment]
The think the box to install the package does need to be ticked. The [API documentation](http://zopeschema.readthedocs.io/en/latest/api.html) is mostly empty because the build [couldn't load the schema package](http://readthedocs.org/projects/zopeschema/builds/4405379/). I believe @tseaver is the doc owner on RTD.

File: docs/conf.py [hunk 1]
```diff
@@ -25,7 +25,13 @@
 
 # Add any Sphinx extension module names here, as strings. They can be extensions
 # coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
-extensions = ['sphinx.ext.autodoc', 'sphinx.ext.doctest', 'sphinx.ext.viewcode']
+extensions = [
+    'sphinx.ext.autodoc',
+    'sphinx.ext.doctest',
+    'sphinx.ext.viewcode',
+    'sphinx.ext.intersphinx',
+    'repoze.sphinx.autointerface',
```
Comment (tseaver) [inline docs/conf.py]: @jamadden This change caused `tox -e docs` to break, as it was not installing `repoze.sphinx.autointerface`.  I fixed that in 357914bf7b15081caff5970711ac043cce10bb2c.

Comment (jamadden) [inline docs/conf.py]: Snap. I'm sorry. I didn't even think to check tox.

Comment (jamadden) [inline docs/conf.py]: Maybe these things should be in the `docs` extra that already includes Sphinx? And then tox could install `.[docs]`.

**Comment by tseaver** [comment]
This PR also broke `tox -e docs` because it pulled in a bunch of previously-unchecked doctest examples:

```python
Document: api
-------------
**********************************************************************
File "api.rst", line 19, in default
Failed example:
    f = Field()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f = Field()
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 20, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 22, in default
Failed example:
    f = Field(title=u('sample'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 23, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 26, in default
Failed example:
    f = Field(title=u('sample'), description=u('blah blah\nblah'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'), description=u('blah blah\nblah'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 27, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 21, in default
Failed example:
    f = Field()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f = Field()
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 22, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 24, in default
Failed example:
    f = Field(title=u('sample'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 25, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 28, in default
Failed example:
    f = Field(title=u('sample'), description=u('blah blah\nblah'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'), description=u('blah blah\nblah'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 29, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 14, in default
Failed example:
    b = Bool()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        b = Bool()
    NameError: name 'Bool' is not defined
**********************************************************************
File "api.rst", line 15, in default
Failed example:
    IFromUnicode.providedBy(b)
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        IFromUnicode.providedBy(b)
    NameError: name 'IFromUnicode' is not defined
**********************************************************************
File "api.rst", line 17, in default
Failed example:
    b.fromUnicode('True')
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[3]>", line 1, in <module>
        b.fromUnicode('True')
    AttributeError: 'function' object has no attribute 'fromUnicode'
**********************************************************************
File "api.rst", line 19, in default
Failed example:
    b.fromUnicode('')
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[4]>", line 1, in <module>
        b.fromUnicode('')
    AttributeError: 'function' object has no attribute 'fromUnicode'
**********************************************************************
File "api.rst", line 21, in default
Failed example:
    b.fromUnicode('true')
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[5]>", line 1, in <module>
        b.fromUnicode('true')
    AttributeError: 'function' object has no attribute 'fromUnicode'
**********************************************************************
File "api.rst", line 23, in default
Failed example:
    b.fromUnicode('false') or b.fromUnicode('False')
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[6]>", line 1, in <module>
        b.fromUnicode('false') or b.fromUnicode('False')
    AttributeError: 'function' object has no attribute 'fromUnicode'
**********************************************************************
File "api.rst", line 19, in default
Failed example:
    f = Field()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f = Field()
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 20, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 22, in default
Failed example:
    f = Field(title=u('sample'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 23, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 26, in default
Failed example:
    f = Field(title=u('sample'), description=u('blah blah\nblah'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'), description=u('blah blah\nblah'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 27, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 8, in default
Failed example:
    f = Int()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Int()
    NameError: name 'Int' is not defined
**********************************************************************
File "api.rst", line 9, in default
Failed example:
    f.fromUnicode("125")
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.fromUnicode("125")
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 11, in default
Failed example:
    f.fromUnicode("125.6") #doctest: +IGNORE_EXCEPTION_DETAIL
Expected:
    Traceback (most recent call last):
    ...
    ValueError: invalid literal for int(): 125.6
Got:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        f.fromUnicode("125.6") #doctest: +IGNORE_EXCEPTION_DETAIL
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 22, in default
Failed example:
    f = Field()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f = Field()
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 23, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 25, in default
Failed example:
    f = Field(title=u('sample'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 26, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 29, in default
Failed example:
    f = Field(title=u('sample'), description=u('blah blah\nblah'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'), description=u('blah blah\nblah'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 30, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 19, in default
Failed example:
    f = Field()
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f = Field()
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 20, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 22, in default
Failed example:
    f = Field(title=u('sample'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 23, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 26, in default
Failed example:
    f = Field(title=u('sample'), description=u('blah blah\nblah'))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[0]>", line 1, in <module>
        f = Field(title=u('sample'), description=u('blah blah\nblah'))
    NameError: name 'Field' is not defined
**********************************************************************
File "api.rst", line 27, in default
Failed example:
    f.__doc__, f.title, f.description
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[1]>", line 1, in <module>
        f.__doc__, f.title, f.description
    NameError: name 'f' is not defined
**********************************************************************
File "api.rst", line 16, in default
Failed example:
    t = Text(constraint=lambda v: 'x' in v)
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[2]>", line 1, in <module>
        t = Text(constraint=lambda v: 'x' in v)
    NameError: name 'Text' is not defined
**********************************************************************
File "api.rst", line 17, in default
Failed example:
    t.fromUnicode(b("foo x spam"))
Expected:
    Traceback (most recent call last):
    ...
    WrongType: ('foo x spam', <type 'unicode'>, '')
Got:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[3]>", line 1, in <module>
        t.fromUnicode(b("foo x spam"))
    NameError: name 't' is not defined
**********************************************************************
File "api.rst", line 21, in default
Failed example:
    t.fromUnicode(u("foo x spam"))
Exception raised:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[4]>", line 1, in <module>
        t.fromUnicode(u("foo x spam"))
    NameError: name 't' is not defined
**********************************************************************
File "api.rst", line 23, in default
Failed example:
    t.fromUnicode(u("foo spam"))
Expected:
    Traceback (most recent call last):
    ...
    ConstraintNotSatisfied: (u'foo spam', '')
Got:
    Traceback (most recent call last):
      File "/opt/Python-2.7.11/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest default[5]>", line 1, in <module>
        t.fromUnicode(u("foo spam"))
    NameError: name 't' is not defined
**********************************************************************
1 items had failures:
  43 of  51 in default
51 tests in 1 items.
8 passed and 43 failed.
***Test Failed*** 43 failures.
```

**Comment by jamadden** [comment]
Then IMHO, it seems like that should either be dropped from tox.ini, or fixed and added to travis. I'm happy to submit a PR that does either one. Let me know if you prefer one of those options or another one.

**Comment by tseaver** [comment]
#26 switches `.travis.yml` to use tox, enabling `coverage` and `docs` environments.

**Comment by jamadden** [comment]
Sorry again for the tox breakage. Thanks for the help turning it on in Travis. Would it be possible to enable installing the package in the virtualenv? The API page is still blank.

**Comment by tseaver** [comment]
@jamadden Thanks for the doctest fixes.  I just tweaked the RTD config to use a virtualenv:  we'll have to see if that fixes the issue (`repoze.sphinx.autointerface` was already in the `rtd.txt` requirements file).

**Comment by tseaver** [comment]
Yay!  Looks like [it did](http://zopeschema.readthedocs.io/en/latest/api.html).

**Comment by jamadden** [comment]
Thank you! Now the `z3c.schema` [docs](http://z3cschema.readthedocs.io/en/latest/dateselect.html) have something to link to!

**Pull request merged**